File: src/java/PhantasyIslandJavaRemoteControl/PhantasyIslandJavaRemoteControl/radio/WaitToken.java
package PhantasyIslandJavaRemoteControl.radio;

import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;
import java.util.function.Function;

/**
 * 轻量级的请求-响应等待令牌。
 */
public class WaitToken<T> {
    public final String waitCmd;
    public final long timeBaseId;
    private Map<String, Object> response = null;
    private T processedResponse = null;
    private final CountDownLatch latch = new CountDownLatch(1);
    private final CompletableFuture<T> future = new CompletableFuture<>();
    private Function<Map<String, Object>, T> postProcessor = null;

    public WaitToken(String waitCmd, long timeBaseId) {
        this.waitCmd = waitCmd;
        this.timeBaseId = timeBaseId;
    }

    public boolean isDone() {
        return latch.getCount() == 0;
    }

    public WaitToken<T> setPostProcessor(Function<Map<String, Object>, T> processor) {
        this.postProcessor = processor;
        return this;
    }

    @SuppressWarnings("unchecked")
    private T applyPostProcessor(Map<String, Object> data) {
        T result;
        if (postProcessor != null) {
            result = postProcessor.apply(data);
        } else {
            result = (T) data;
        }
        this.processedResponse = result;
        return result;
    }

    /**
     * 填充响应并唤醒所有等待者。
     */
    public void complete(Map<String, Object> data) {
        this.response = data;
        T result = applyPostProcessor(data);
        latch.countDown();
        future.complete(result);
    }

    /**
     * 获取原始响应数据。
     */
    public Map<String, Object> getResponse() {
        return response;
    }

    /**
     * 获取后处理后的响应数据。
     */
    public T getProcessedResponse() {
        return processedResponse;
    }

    /**
     * 同步等待响应。
     */
    public T waitResponse(double timeoutSeconds) {
        try {
            if (latch.await((long) (timeoutSeconds * 1000), TimeUnit.MILLISECONDS)) {
                return processedResponse;
            }
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
        return null;
    }

    /**
     * 获取用于异步等待的 CompletableFuture。
     */
    public CompletableFuture<T> asFuture() {
        return future;
    }

    @Override
    public String toString() {
        String status = isDone() ? "done" : "pending";
        return String.format("<WaitToken cmd=%s %s>", waitCmd, status);
    }
}

File: src/kt/PhantasyIslandKotlinRemoteControl/PhantasyIslandKotlinRemoteControl/radio/WaitToken.kt
package PhantasyIslandKotlinRemoteControl.radio

import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit
import java.util.concurrent.CompletableFuture

/**
 * 轻量级的请求-响应等待令牌。
 */
class WaitToken<T>(
    val waitCmd: String,
    val timeBaseId: Int
) {
    var response: T? = null
    var processedResponse: Any? = null
    private val latch = CountDownLatch(1)
    private val future = CompletableFuture<Any?>()
    private var postProcessor: ((T) -> Any)? = null

    /** 令牌是否已收到响应。 */
    val isDone: Boolean
        get() = latch.count == 0L

    /**
     * 注册后处理回调。当收到 response 后，会自动调用此回调进行数据格式转换。
     */
    fun setPostProcessor(processor: (T) -> Any): WaitToken<T> {
        this.postProcessor = processor
        return this
    }

    private fun applyPostProcessor(data: T): Any? {
        val result = postProcessor?.invoke(data) ?: data
        this.processedResponse = result
        return result
    }

    /**
     * 填充响应并唤醒所有等待者。
     */
    fun complete(data: T) {
        this.response = data
        val result = applyPostProcessor(data)
        latch.countDown()
        future.complete(result)
    }

    /**
     * 阻塞当前线程直到收到响应或超时。
     * @param timeoutSec 超时秒数
     * @return 后处理结果，超时返回 null
     */
    fun wait(timeoutSec: Double = 3.0): Any? {
        val success = latch.await((timeoutSec * 1000).toLong(), TimeUnit.MILLISECONDS)
        return if (success) processedResponse else null
    }

    /**
     * 获取用于异步等待的 CompletableFuture。
     */
    fun asFuture(): CompletableFuture<Any?> {
        return future
    }

    override fun toString(): String {
        val status = if (isDone) "done" else "pending"
        return "<WaitToken cmd=$waitCmd $status>"
    }
}
